src/main/user-impl/java/com/mysql/cj/jdbc/ha/ca/plugins/failover/FailoverConnectionPlugin.java

```diff
@@ -0,0 +1,1289 @@
+
+    if (this.gatherPerfMetricsSetting) {
+      long currentTimeMs = System.currentTimeMillis();
+      this.metrics.registerWriterFailoverProcedureTime(
+          currentTimeMs - this.failoverStartTimeMs);
+      this.failoverStartTimeMs = 0;
+    }
+
+    if (failoverResult == null || !failoverResult.isConnected()) {
+      // "Unable to establish SQL connection to writer node"
+      processFailoverFailure(Messages.getString("ClusterAwareConnectionProxy.2"));
+      return;
+    }
+
+    if (!Util.isNullOrEmpty(failoverResult.getTopology())) {
+      this.hosts = failoverResult.getTopology();
+    }
+
+    if (this.gatherPerfMetricsSetting) {
+      this.metrics.registerFailoverConnects(true);
+    }
+
+    // successfully re-connected to the same writer node
+    setConnectionProxy(updateCurrentConnection(
+        failoverResult.getNewConnection(),
+        WRITER_CONNECTION_INDEX));
+
+    this.logger.logDebug(
+        Messages.getString(
+            "ClusterAwareConnectionProxy.15",
+            new Object[] {this.hosts.get(this.currentHostIndex)}));
+  }
+
+  /**
+   * Get this connection's proxy.
+   * A multi-host connection may not be at top level in the multi-host connections chain. In such case the first connection in the chain is available as a
+   * proxy.
+   *
+   * @return
+   *         Returns this connection's proxy if there is one or itself if this is the first one.
+   */
+  protected JdbcConnection getProxy() {
+    return this.topProxyConnection != null ? this.topProxyConnection
+        : this.thisAsConnection;
+  }
+
+  protected void invalidateCurrentConnection() {
+    final JdbcConnection conn = this.currentConnectionProvider.getCurrentConnection();
+    if (this.inTransaction) {
+      try {
+        conn.rollback();
+      } catch (SQLException e) {
+        // eat
+      }
+    }
+
+    try {
+      if (conn != null && !conn.isClosed()) {
+        conn.realClose(true, !conn.getAutoCommit(), true, null);
+      }
+    } catch (SQLException e) {
+      // swallow this exception, current connection should be useless anyway.
+    }
+  }
+
+  protected synchronized void pickNewConnection() throws SQLException {
+    if (this.isClosed && this.closedExplicitly) {
+      this.logger.logDebug(Messages.getString("ClusterAwareConnectionProxy.1"));
+
+      return;
+    }
+
+    if (isConnected() || Util.isNullOrEmpty(this.hosts)) {
+      failover(this.currentHostIndex);
+      return;
+    }
+
+    if (shouldAttemptReaderConnection()) {
+      failoverReader(NO_CONNECTION_INDEX);
+      return;
+    }
+
+    try {
+      connectTo(WRITER_CONNECTION_INDEX);
+      if (isExplicitlyReadOnly() && this.currentHostIndex != NO_CONNECTION_INDEX) {
+        topologyService.setLastUsedReaderHost(this.hosts.get(this.currentHostIndex));
+      }
+    } catch (SQLException e) {
+      failover(WRITER_CONNECTION_INDEX);
+    }
+  }
+
+  protected void setConnectionProxy(JdbcConnection conn) {
```

Comment:
It might be better to leave this method in connection proxy class.

Reply:
As discussed offline, we don't need setProxy at all.